File: server/SpatialGrid.js
/**
 * High-performance 1D Flat Spatial Grid for zero-allocation broadphase collision checks.
 */
class SpatialGrid {
  constructor(worldWidth, worldHeight, cellSize = 400) {
    this.worldWidth = worldWidth;
    this.worldHeight = worldHeight;
    this.cellSize = cellSize;
    this.cols = Math.ceil(worldWidth / cellSize);
    this.rows = Math.ceil(worldHeight / cellSize);
    this.cellCount = this.cols * this.rows;
    this.cells = new Array(this.cellCount);
    for (let i = 0; i < this.cellCount; i++) {
      this.cells[i] = [];
    }
  }

  _getCellIndex(cx, cy) {
    return cy * this.cols + cx;
  }

  _getCellCoords(x, y) {
    const cx = Math.max(0, Math.min(this.cols - 1, (x / this.cellSize) | 0));
    const cy = Math.max(0, Math.min(this.rows - 1, (y / this.cellSize) | 0));
    return { cx, cy };
  }

  clear() {
    for (let i = 0; i < this.cellCount; i++) {
      this.cells[i].length = 0;
    }
  }

  insert(entity) {
    const cx = Math.max(0, Math.min(this.cols - 1, (entity.x / this.cellSize) | 0));
    const cy = Math.max(0, Math.min(this.rows - 1, (entity.y / this.cellSize) | 0));
    this.cells[cy * this.cols + cx].push(entity);
  }

  getNearby(x, y, radius = 400) {
    const minCx = Math.max(0, Math.min(this.cols - 1, ((x - radius) / this.cellSize) | 0));
    const minCy = Math.max(0, Math.min(this.rows - 1, ((y - radius) / this.cellSize) | 0));
    const maxCx = Math.max(0, Math.min(this.cols - 1, ((x + radius) / this.cellSize) | 0));
    const maxCy = Math.max(0, Math.min(this.rows - 1, ((y + radius) / this.cellSize) | 0));
    const results = [];

    for (let cy = minCy; cy <= maxCy; cy++) {
      const rowOffset = cy * this.cols;
      for (let cx = minCx; cx <= maxCx; cx++) {
        const list = this.cells[rowOffset + cx];
        for (let i = 0; i < list.length; i++) {
          results.push(list[i]);
        }
      }
    }
    return results;
  }
}

module.exports = SpatialGrid;

File: server_go/game/spatial_grid.go
package game

import "math"

type SpatialGrid struct {
	WorldWidth  float64
	WorldHeight float64
	CellSize    float64
	Cols        int
	Rows        int
	CellCount   int
	Cells       [][]*Player
}

func NewSpatialGrid(worldWidth, worldHeight, cellSize float64) *SpatialGrid {
	cols := int(math.Ceil(worldWidth / cellSize))
	rows := int(math.Ceil(worldHeight / cellSize))
	cellCount := cols * rows

	cells := make([][]*Player, cellCount)
	for i := range cells {
		cells[i] = make([]*Player, 0, 16)
	}

	return &SpatialGrid{
		WorldWidth:  worldWidth,
		WorldHeight: worldHeight,
		CellSize:    cellSize,
		Cols:        cols,
		Rows:        rows,
		CellCount:   cellCount,
		Cells:       cells,
	}
}

func (g *SpatialGrid) Clear() {
	for i := range g.Cells {
		g.Cells[i] = g.Cells[i][:0]
	}
}

func (g *SpatialGrid) Insert(p *Player) {
	cx := int(p.X / g.CellSize)
	cy := int(p.Y / g.CellSize)

	if cx < 0 {
		cx = 0
	} else if cx >= g.Cols {
		cx = g.Cols - 1
	}

	if cy < 0 {
		cy = 0
	} else if cy >= g.Rows {
		cy = g.Rows - 1
	}

	idx := cy*g.Cols + cx
	g.Cells[idx] = append(g.Cells[idx], p)
}

func (g *SpatialGrid) GetNearby(x, y, radius float64) []*Player {
	minCx := int((x - radius) / g.CellSize)
	maxCx := int((x + radius) / g.CellSize)
	minCy := int((y - radius) / g.CellSize)
	maxCy := int((y + radius) / g.CellSize)

	if minCx < 0 {
		minCx = 0
	}
	if maxCx >= g.Cols {
		maxCx = g.Cols - 1
	}
	if minCy < 0 {
		minCy = 0
	}
	if maxCy >= g.Rows {
		maxCy = g.Rows - 1
	}

	results := make([]*Player, 0, 32)
	for cy := minCy; cy <= maxCy; cy++ {
		rowOffset := cy * g.Cols
		for cx := minCx; cx <= maxCx; cx++ {
			list := g.Cells[rowOffset+cx]
			results = append(results, list...)
		}
	}
	return results
}
